Add De Lijn (Flemish Public Transport) component

bollewolle opened the pull request:
## Breaking Change:

none

## Description:
Add [De Lijn](https://www.delijn.be) (the Flemish Public Transport agency) support. The use of the API's of De Lijn does require a subscription key that needs to be added to the configuration. You can register for a free developer account at https://data.delijn.be/ where you can find the subscription key.

Works best with the following custom lovelace card: https://github.com/bollewolle/delijn-card

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/9507

## Example entry for `configuration.yaml`:
```yaml
sensor:
  # De Lijn public transport
  - platform: delijn
    sub_key: 'abcdefg'
    nextpassage:
    - stop_id: '200018'
      max_passages: 5
    - stop_id: '201169'
      max_passages: 20
```
Replace the abcdefg with the API subscription key of De Lijn.
Stop Id's are 6 digit number that can be found at each physical stop of De Lijn in Flanders, but can also be found on their site here: https://www.delijn.be/en/haltes/

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

bollewolle committed:
Initial commit of De Lijn (Flemish Public Transport) component

fabaff commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+    @property
+    def device_state_attributes(self):
+        """Return attributes for the sensor."""
+        return self._attributes
```
Please include an attribution for De Lijn.

bollewolle replied:
@fabaff Could you give some more information about your request? I basically used the already existing integration of Ruter (https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/ruter/sensor.py) as basis where the above piece of code is exactly the same. But this kind of coding is quite new to me so all help is welcome if something else is needed here.

MartinHjelmare replied:
See eg entur public transport:
https://github.com/home-assistant/home-assistant/blob/a30c37017b7782473294d7999e85d7a369a0539a/homeassistant/components/entur_public_transport/sensor.py#L172

bollewolle replied:
Changes made in last commit

fabaff commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+        """Get the latest data from the De Lijn API."""
+        await self.line.get_passages()
+        if self.line.passages is None:
+            _LOGGER.error("No data recieved from De Lijn.")
```
Please remove the period at the end of the log message to be in sync with all other log messages.

bollewolle replied:
Corrected in local version, will push when others comments are done.

bollewolle replied:
Practical question: should I mark the "Resolve conversation" myself or is this expected to be done by one of the HA core developpers? First time I do this so need to learn it some way :). Thanks!

MartinHjelmare replied:
Either way is fine. If you think it's resolved you can mark it resolved.

fabaff commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+DEFAULT_NAME = 'De Lijn'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SUB_KEY): cv.string,
```
Please use `CONF_API_KEY`.

bollewolle replied:
Corrected in local version, will push when others comments are done.

fabaff commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+from homeassistant.helpers.entity import Entity
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.exceptions import PlatformNotReady
```
Please keep the imports [ordered](https://developers.home-assistant.io/docs/en/development_guidelines.html#ordering-of-imports).

bollewolle replied:
Corrected in local version by running isort, will push when others comments are done.

fabaff commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SUB_KEY): cv.string,
+    vol.Required(CONF_NEXT_PASSAGE): [{
```
I suggest that this is renamed to `stations`.

bollewolle replied:
I don't really follow this suggestion if that's ok for you. For me stations are more related to trains, for trams and busses "stops" is the correct terminology (and it's also used like this on the official site of De Lijn: https://www.delijn.be/en/haltes/).
With Passage I mean the next passage of a bus at a stop. Which is also the correct terminology (and also here it's the terminology used on the official site of De Lijn: https://www.delijn.be/en/haltes/halte/200552/Gent_Korenmarkt_perron_2).
Hope that explains the use of this terminology.

MartinHjelmare replied:
Can we rename this to `CONF_NEXT_DEPARTURE` which should hold `next_departure`?

bollewolle replied:
Changes made in last commit

fabaff commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+    vol.Required(CONF_SUB_KEY): cv.string,
+    vol.Required(CONF_NEXT_PASSAGE): [{
+        vol.Required(CONF_STOP_ID): cv.string,
+        vol.Optional(CONF_MAX_PASSAGES, default=5): cv.positive_int}]
```
From my understanding are those the lines which are monitored. Would be something like `CONF_MAX_LINES` make more sense?

bollewolle replied:
I don't really follow this suggestion if that's ok for you. For me stations are more related to trains, for trams and busses "stops" is the correct terminology (and it's also used like this on the official site of De Lijn: https://www.delijn.be/en/haltes/).
With Passage I mean the next passage of a bus at a stop, in this case the default is that it will retrieve the next 5 passages of busses or trams at a stop. Which is also the correct terminology (and also here it's the terminology used on the official site of De Lijn: https://www.delijn.be/en/haltes/halte/200552/Gent_Korenmarkt_perron_2).
Hope that explains the use of this terminology.

MartinHjelmare replied:
Do we have an existing corresponding config option in the other transport integrations, where we call this something else? Please review them and check. We need to make sure that we don't dilute this concept into different names.

bollewolle replied:
Thanks for the feedback.

I quickly went over all the components of the type [transport](https://www.home-assistant.io/components/#transport). This is what I came up with.

[entur_public_transport](https://www.home-assistant.io/components/entur_public_transport/) uses number_of_departures as variable which perhaps is the closes to what I'm using it for. Although personally I don't like "departure" as terminology for a bus or tram stop as it's not really grammatically correct.

[Mvglive](https://www.home-assistant.io/components/mvglive/) uses number as variable for the same purpose.

[Rmvtransport](https://www.home-assistant.io/components/rmvtransport/) uses max_journeys as a variable, but it's quite different from what I'm using it for. It seens that this integration really works with a "from" and "to" to define journeys (I want to go from point A to point B and that's my journey). So in that concept the max_journeys variable makes sense.

So bottom-line, yes it seems that the "number_of_departures" from entur_public_transport and "number" from Mvglive integrations are used at the moment for similar purposes. If you'd really like me to make the adjustment, I would personally vote for  "number_of_departures" as just "number" does not say much. 

Any advise/proposition?

Thanks in advance!

MartinHjelmare replied:
`number_of_departures` sounds good!

bollewolle replied:
Changes made in last commit

bollewolle committed:
Code corrections as per HA dev's requests

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+    tasks = [sensor.async_update() for sensor in sensors]
+    if tasks:
+        await asyncio.wait(tasks)
+    if not all(sensor._attributes for sensor in sensors):
```
We shouldn't access private attributes outside the instance.

I suggest calling line.get_passages directly instead for all lines.

`async_add_entities(..., True)` will make sure `entity.async_update` is called during entity addition.

bollewolle replied:
Changes made in last commit

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+        try:
+            attributes = {}
+            first = self.line.passages[0]
+            self._state = first['due_in_min']
```
We don't want relative times in the state machine. Relative times will cause not needed state updates.

I suggest to calculate the absolute utc iso timestamp and return that as state. Also set the entity property `device_class` to timestamp. That will allow to show the absolute timestamp as relative time in the frontend.

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

bollewolle replied:
Just to be sure I understand correctly. You're proposing to use a timestamp as the self._state, correct? Because the issue of the API is that their datetimes are naive. So I'd have to do some magic to make them into a UTC timestamp. Would this kind of code be ok?

`
local_timezone = pytz.timezone("Europe/Brussels")
if first['due_at_rt'] is not None:
    first_passage_naive = datetime.strptime(first['due_at_rt'], '%Y-%m-%dT%H:%M:%S')
else:
    first_passage_naive = datetime.strptime(first['due_at_sch'], '%Y-%m-%dT%H:%M:%S')
first_passage_local = local_timezone.localize(first_passage_naive)
first_passage_utc = first_passage_local.astimezone(pytz.utc)
self._state = first_passage_utc
`

The localized timestamp would be something like this: 2019-07-24 13:08:31+02:00
The UTC timestamp, that would be used for the self._state, would be something like: 2019-07-24 11:08:31+00:00

Is that what you expect to be used? Because event wit the device_class property set it doesn't seem to do much interpretation in the frontend in my dev environment.

Thanks in advance for the feedback.

MartinHjelmare replied:
The state timestamp needs to be an iso datetime string.

Also, please use our datetime utils as much as possible.
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/util/dt.py

MartinHjelmare replied:
Here are the lovelace config options for timestamp entity for the entity card:
https://www.home-assistant.io/lovelace/entities/#format

bollewolle replied:
Changes applied in last commit. Made use of the dtutils within home-assistant as well.

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+            attributes['line_number_public'] = first['line_number_public']
+            attributes['line_transport_type'] = first['line_transport_type']
+            attributes['final_destination'] = first['final_destination']
+            attributes['due_at_sch'] = first['due_at_sch']
```
What does `due_at_sch` mean?

bollewolle replied:
Sometimes the real-time data is not available (or not made available) by the API, for example when the tram/bus did not leave yet at the starting station there is not real-time info yet but only the scheduled info is available. That's why there's a split made between scheduled (due_at_sch) and real-time (due_at_rt) times.

MartinHjelmare replied:
Can we write out `schedule` instead of `sch`, and real-time` instead of `rt`?

bollewolle replied:
Changed the due_at_sch to due_at_schedule and due_at_rt to due_at_realtime in the attributes[] but not in the first[] as those are the names used by the pydelijn package.

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+            attributes['line_transport_type'] = first['line_transport_type']
+            attributes['final_destination'] = first['final_destination']
+            attributes['due_at_sch'] = first['due_at_sch']
+            attributes['due_at_rt'] = first['due_at_rt']
```
What does this mean?

bollewolle replied:
Sometimes the real-time data is not available (or not made available) by the API, for example when the tram/bus did not leave yet at the starting station there is not real-time info yet but only the scheduled info is available. That's why there's a split made between scheduled (due_at_sch) and real-time (due_at_rt) times.

bollewolle replied:
Changed the due_at_sch to due_at_schedule and due_at_rt to due_at_realtime in the attributes[] but not in the first[] as those are the names used by the pydelijn package.

bollewolle committed:
changes to variable naming, setting attribution and states, plus some smaller optimizations

bollewolle commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None):
+    """Create the sensor."""
+    from pydelijn.api import Passages
```
Changes made in last commit

bollewolle commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,107 @@
+    """Create the sensor."""
+    from pydelijn.api import Passages
+
+    api_key = config.get(CONF_API_KEY)
```
Changes made in last commit

bollewolle committed:
Overlooked some linting issues, these are now fixed

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,135 @@
+})
+
+
+def due_in_minutes(timestamp):
```
This isn't used.

bollewolle replied:
removed this unused leftover in latest commit

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,135 @@
+            attributes['final_destination'] = first['final_destination']
+            attributes['due_at_schedule'] = first['due_at_sch']
+            attributes['due_at_realtime'] = first['due_at_rt']
+            attributes['due_in_min'] = first['due_in_min']
```
Please don't include any relative times in the state attributes.

bollewolle replied:
I've updated the pydelijn package in the latest commit so it has the option to convert the API timestamps of De Lijn (which are always of timzone "Europe/Brussels") to UTC timestamps. This solves the overall issue of the use of relative times.

MartinHjelmare replied:
Please remove due_in_min from the attributes.

bollewolle replied:
Can you please give some more feedback as to why? Because that's the exact purpose of why I built this component, to reflect in how much time the next bus/tram will pass by at a certain stop. Are you saying the difference in time should be something that's calculated in Javascript in custom Lovelace cards then instead? Just to understand the reasoning and be sure what next action I need to make on my behalf. Thanks.

MartinHjelmare replied:
Yes. The state which is now an absolute utc timestamp, can be shown as relative time in the lovelace card. We don't allow relative times in state or state attributes, in the state machine.

bollewolle replied:
OK, removed in the latest commit. Will use some logic in the custom Lovelace card instead.

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,135 @@
+    @property
+    def device_state_attributes(self):
+        """Return attributes for the sensor."""
+        self._attributes[ATTR_ATTRIBUTION] = ATTRIBUTION
```
Include the attribution already when creating the dict instance attribute in init method.

bollewolle replied:
Added the attribution to the init method in the latest commit.

MartinHjelmare replied:
We can remove this if we update instead of replacing the attributes.

bollewolle replied:
Done, updated in the latest commit

bollewolle committed:
Updated pydelijn version requirement to 0.5.1 so UTC timestamps can be used instead of relative/local times.
Removed unused definition

bollewolle committed:
Updated pydelijn version requirement to 0.5.1 in requirements_all.txt

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,123 @@
+                first_passage = first['due_at_realtime']
+            else:
+                first_passage = first['due_at_schedule']
+            self._state = first_passage
```
Is this an iso formatted datetime string?

bollewolle replied:
Yes they are, please see red box.
![image](https://user-images.githubusercontent.com/368306/61940893-a33ab680-af96-11e9-9ffd-28f20540453e.png)

All dates are now iso formatted (see green boxes as well)

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,123 @@
+            attributes['due_at_realtime'] = first['due_at_realtime']
+            attributes['due_in_min'] = first['due_in_min']
+            attributes['next_passages'] = self.line.passages
+            self._attributes = attributes
```
Update the instance attribute dict with the new attributes instead of replacing it.

bollewolle replied:
Just to be sure I understand you correctly:
I need to replace all the attributes['some_key'] with self.attributes['some_key'] instead of creating a dict and then assigning (and thus replacing) it to the self._attributes. Is that understanding correct?

bollewolle replied:
OK, I've updated this in the latest commit (as per my description above).

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,123 @@
+        self._attributes = {}
+        self._name = name
+        self._state = None
+        self._attributes[ATTR_ATTRIBUTION] = ATTRIBUTION
```
Include the attribution when creating the dict.

bollewolle replied:
Updated in the latest commit

bollewolle committed:
Update the self._attributes dict directly instead of replacing it
Assign ATTRIBUTION while creating the _attributes dict
Remove the ATTRIBUTION assignment in device_state_attributes as it's updated in the async_update now.

bollewolle committed:
Linting issue (lenght of 2 lines) solved

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,122 @@
+            self._attributes['due_at_realtime'] = first['due_at_realtime']
+            self._attributes['due_in_min'] = first['due_in_min']
+            self._attributes['next_passages'] = self.line.passages
+            self._attributes[ATTR_ATTRIBUTION] = ATTRIBUTION
```
We don't need to update the attribution. It's constant.

bollewolle replied:
OK, removed in the latest commit. Forgot to remove this one before.

MartinHjelmare commented on homeassistant/components/delijn/sensor.py:
```diff
@@ -0,0 +1,122 @@
+            self._attributes['final_destination'] = first['final_destination']
+            self._attributes['due_at_schedule'] = first['due_at_schedule']
+            self._attributes['due_at_realtime'] = first['due_at_realtime']
+            self._attributes['due_in_min'] = first['due_in_min']
```
Please remove this.

bollewolle replied:
OK, removed in the latest commit. Will use some logic in the custom Lovelace card instead.

MartinHjelmare replied:
https://www.home-assistant.io/lovelace/entities/#format

bollewolle committed:
Removed a relative time attribute
Updated a linting issue in the LOGGER (used % instead of the format)

MartinHjelmare commented:
Great!
